core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -723,14 +726,14 @@ class GroupCoordinator(val brokerId: Int,
         responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.COORDINATOR_NOT_AVAILABLE })
       } else if (group.isStaticMemberFenced(memberId, groupInstanceId)) {
         responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.FENCED_INSTANCE_ID })
-      } else if ((generationId < 0 && group.is(Empty)) || (producerId != NO_PRODUCER_ID)) {
+      } else if (generationId >= 0 && generationId != group.generationId) {
+        // Validate non-zero generation id for both transactional and non-transactional commits.
+        responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.ILLEGAL_GENERATION })
+      } else if (group.isUnknownCommit(memberId, producerId, generationId)) {
+        responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.UNKNOWN_MEMBER_ID })
+      } else if (group.isManualCommit(generationId) || producerId != NO_PRODUCER_ID) {
```

Comment:
prop: I'd propose we just make the `producerId != NO_PRODUCER_ID` condition in the first place to explicitly separate the checking logic for transactional / non-txnal commit starting at 732, e.g.:

```
if (producerId != NO_PRODUCER_ID) {
   // check if it is unknown
} else {
   // check if it is unknown, and
   // check if it is manual
}
```

Because if `producer != NO_PRODUCER_ID` it should never be manual, or am I wrong?

Reply:
I think the example with KIP-98 is actually doing assign mode, single writer producer.
So the answer is false, txn commit could be manual

Reply:
What I'm thinking is mainly for debuggability, wondering if we explicitly separate the logic of txn / non-txn commit it is more readable -- currently the `isUnknownCommit` and `isManualCommit` are a bit mixed. 

Thinking about this a bit more, how about the following within a single function (ordered):

1) if generation id < 0 or state is Empty, we just bindly store offset (we can actually be stricter and check that member.id should always be UNKNOWN, but this is not there before so we do not really need to enforce it).

2) else, check memberId != JoinGroupRequest.UNKNOWN_MEMBER_ID && !has(memberId).

For both txn commit or consumer commit, 2) should be the case. So we do not need to really distinguish on producerId --- i.e. previously if the producerId is set we do not do the check of 2), now we always do it regardless. Is that correct?

Reply:
I think this actually looks like a much simpler version, just to add up after case 2), we need two more cases:
3. else if generationId >= 0 && generationId != group.generationId we return ILLEGAL_GENERATION
4. else if producerId != NO_PRODUCER_ID we blindly store offsets

The case 3 is necessary for generation check on both txn and non-txn commits. The case 4) is necessary as we don't want txn commit fall into group case match, where we actually do a `val member = group.get(memberId)`. The member.id could still be unknown at this moment with an older txn producer who doesn't pack in consumer metadata.

Reply:
If we refactor the logic in the proposed way, `testCommitOffsetInCompletingRebalanceFromUnknownMemberId` will fail as we are not able to catch the unknown member.id when the commit is non-txn.

My main argument is that the complexity we are handling here is under-estimated, in a sense that for txn commit we also need to worry about whether this producer is on the old version or new version. I would attempt a separation logic just as what we did for `doJoinGroup` and `doUnknownJoinGroup` which is able to separate the txn vs non-txn commits